### ISSUE TITLE
Question: creating or keeping the lock instance

### ISSUE THREAD

**Issue by ffeu** [tekartik/synchronized.dart#4]
Should I create the Lock on every execution, as the examples demonstrate, or, if I'm willing to execute it often, should I keep that instance and reuse it?

Would that be wrong?

```
class MyClass {
  static Lock _lock = new Lock();
  static Future<void> myMethod() async {
    await _lock.synchronized(() async {
      ...
    });
  }
}
```

**Comment by alextekartik** [comment]
Not sure if I understand correctly. In your example (where `_lock` is static and could be final), the instance is created only once during the whole program execution. Are you asking whether your `_lock` should be static or not? if that's the questions, it depends what you want to protect concurrent access from.

If you're fetching and/or accessing global data, a `Lock` could be a global or a static variable.

If you're manipulating data relative to one instance, a `Lock` object should be a member of your class.

Creating a `Lock` object is almost a no-op operation. 

Sorry if my response does not answer what you ask for. In that case can you clarify with a concrete example? thanks

**Comment by ffeu** [comment]
Thanks for your response and sorry - let me rephrase. :)

Considering that I have a `myMethod()` that is bound to asynchronous events and might be executed concurrently, and that it executes 3 steps that must be executed sequentially (just as the example with 11223344 and 12341234).

How does these two snippets differ? Is one of them correct or are they both correct?

### 1) keeping _lock through calls to myMethod()
```
class MyClass {
  static Lock _lock = new Lock();

  static Future<void> myMethod() async {
    await _lock.synchronized(() async {
      step1();
      step2();
      step3();
    });
  }
}
```
### 2) creating a new lock for each call
```
class MyClass {
  static Future<void> myMethod() async {
    Lock lock = new Lock();
    await lock.synchronized(() async {
      step1();
      step2();
      step3();
    });
  }
}

```

**Comment by alextekartik** [comment]
Ok I understand. The second snippet is definitely wrong, since you create a new lock each time so it is not locked the first time you call synchronized and here you only call it once (per object) even if you call `myMethod` multiple times.

The lock object must be shared by the 2 callers in order to work. The first snippet is correct. calling myMethod a second time will block until the first time call is done

I guess my documentation (if any) is not clear on that...

**Comment by ffeu** [comment]
I'm glad I asked, then! :)

In fact, if there were no docs I would '1' directly, but somehow reading the docs I got a bit doubtful. :)
